File: src/problems/leetcode-cn/3637. 三段式数组 I/solution/go/solution.go
func isTrionic(nums []int) bool {
    start, length := 0, len(nums)

    // 少于 4 个不能构成完整坡道
    if length < 4 {
        return false
    }

    // 爬坡闭包函数 (模拟)
    climb := func(cmp func(a, b int) bool) bool {
        origin, end := start, length - 1
        for ; start < end && cmp(nums[start+1], nums[start]); start++ {}
        return start > origin
    }

    if !climb(func(a, b int) bool { return a > b }) ||
    !climb(func(a, b int) bool { return a < b }) ||
    !climb(func(a, b int) bool { return a > b }) {
        return false
    }

    return start == length-1
}

File: src/problems/leetcode-cn/3637. 三段式数组 I/solution/rust/solution.rs
impl Solution {
    pub fn is_trionic(nums: Vec<i32>) -> bool {
        if nums.len() < 4 {
            return false;
        }

        fn climb(nums: &[i32], mut i: usize, cmp: fn(i32, i32) -> bool) -> Option<usize> {
            let start = i;
            while i + 1 < nums.len() && cmp(nums[i + 1], nums[i]) {
                i += 1;
            }
            if i > start { Some(i) } else { None }
        }

        let mut i = 0;

        i = match climb(&nums, i, |a, b| a > b) { Some(x) => x, None => return false };
        i = match climb(&nums, i, |a, b| a < b) { Some(x) => x, None => return false };
        i = match climb(&nums, i, |a, b| a > b) { Some(x) => x, None => return false };

        i == nums.len() - 1
    }
}